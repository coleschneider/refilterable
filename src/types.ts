import { History } from 'history';
import { MutableRefObject } from 'react';

export const CREATE_FILTER_MARKER = Symbol("createFilter");

export interface FilterConfig<T = string> {
	parse?(input: string): T; 
  format?(value: T): string; 
  validate?(input: string, parse: (input: string) => T): boolean;
  defaultValue?: T | undefined;
}

export interface FilterObject<T = string> extends FilterConfig<T> {
  paramName: string;
  parse(input: string): T; 
  format(value: T): string; 
  validate(input: string, parse: (input: string) => T): boolean;
}

/**
 * User defined type guard to check if the passed object is a valid instance of FiterObject
 * 
 * @param filter 
 */
export function isFilterObject<T>(filter: any): filter is FilterObject<T> {
  return filter[CREATE_FILTER_MARKER];
}

export type FiltersContextValue = {
  locationObserver: MutableRefObject<{ 
    watch(paramName: string, callback: Function): Function;
    getParamInfo(paramName: string): { hasParam: boolean, paramValue: any };
    getCurrentParams(): URLSearchParams;
  }>;
  history?: History;
  filterRegistry: Map<string, FilterObject<any>>,
}

export type ForwardHistoryAction = 'PUSH' | 'REPLACE';

export type SetFilterOptions = {
  dry?: boolean;
  action?: ForwardHistoryAction;
};

export type FilterSetter<T = string> = (nextValue: T, options?: SetFilterOptions) => string;
