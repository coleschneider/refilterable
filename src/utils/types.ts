import { History } from 'history';
import { CREATE_FILTER_MARKER } from './constants';

export type ParseFunction<T> = (input: string | undefined) => T | undefined;



export interface FilterConfig<T = string> {
	parse?: ParseFunction<T>;
  format?(value: T): string; 
  validate?(input: string, parse: ParseFunction<T>): boolean;
  defaultValue?: string | undefined;
}

export interface FilterObject<T = string> extends FilterConfig<T> {
  paramName: string;
	parse: ParseFunction<T>;
  format(value: T): string; 
  validate(input: string, parse: ParseFunction<T>): boolean;
}

/**
 * User defined type guard to check if the passed object is a valid instance of FiterObject
 * 
 * @param filter 
 */
export function isFilterObject<T>(filter: any): filter is FilterObject<T> {
  return Boolean(filter[CREATE_FILTER_MARKER]);
}

export type FiltersContextValue = {
  locationObserver: { 
    watch(paramName: string, callback: Function): Function;
    getParamInfo(paramName: string): { hasParam: boolean, paramValue: any };
    getCurrentParams(): URLSearchParams;
  };
  history?: History;
  filterRegistry: Map<string, FilterObject<any>>,
}

export type ForwardHistoryAction = 'PUSH' | 'REPLACE';

export type SetFilterOptions = {
  dry?: boolean;
  action?: ForwardHistoryAction;
};

export type FilterSetter<T = string> = (nextValue: T, options?: SetFilterOptions) => string;
export type FilterResetter = (options?: SetFilterOptions) => string;
