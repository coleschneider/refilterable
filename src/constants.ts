import { SetFilterOptions, ForwardHistoryAction } from './types';

export const defaultSetFilterOptions: SetFilterOptions = Object.freeze({
  dry: false,
  action: 'PUSH' as ForwardHistoryAction,
});