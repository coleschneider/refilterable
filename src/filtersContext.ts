import { createContext } from "react";
import { FiltersContextValue } from './types';
import createLocationObserver from "./createLocationObserver";

export default createContext<FiltersContextValue | null>(null);