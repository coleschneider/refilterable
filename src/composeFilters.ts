import { FILTER_COMPOSITION_MARKER } from './utils/constants';
import { FilterObject, FilterComposition } from './utils/types';

// @todo: support filter compositions in filters
export default function composeFilters(filters: FilterObject[], validate?: (input: any) => boolean): FilterComposition {
  return ({
    filters,
    validate: typeof validate === 'function' ? validate : () => true,
    [FILTER_COMPOSITION_MARKER]: true,
  } as FilterComposition) ;
}