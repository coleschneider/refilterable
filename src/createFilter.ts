import invariant from 'invariant';
import { FilterConfig, FilterObject } from './utils/types';
import { CREATE_FILTER_MARKER } from './utils/constants';


export default function createFilter<T>(paramName: string, config?: FilterConfig<T>): FilterObject<T> {
  invariant(
    String(paramName || '').length > 0,
    `re-filter: param name cannot be empty`,
  );

  const defaultConfig = {
    parse(input: string): any { return input; }, 
    format(value: T): string { return String(value); },
    validate(): boolean { return true; },
    defaultValue: undefined,
    [CREATE_FILTER_MARKER]: true,
  };

  return {
    ...defaultConfig,
    ...config,
    paramName,
  };
}; 