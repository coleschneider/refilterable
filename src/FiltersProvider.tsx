import React, { useEffect, useRef, FunctionComponent } from 'react';
import invariant from 'invariant';
import { History, Location } from 'history';
import filtersContext from './filtersContext';
import { FiltersContextValue, FilterObject } from './types';
import createLocationObserver from './createLocationObserver';
 
type Props = {
  history: History;
  children: JSX.Element[];
}

const FiltersProvider: FunctionComponent<Props> = ({ history, children }: Props) => {
  invariant(
    history,
    `re-filter: FiltersProvider was not passed a history instance. 
     Make sure you pass it via the history prop`,
  );

  const filterRegistry = useRef<Map<string, FilterObject<any>>>(new Map());
  const locationObserver = useRef(createLocationObserver(history.location.search));

  useEffect(() => {
    // listen to location changes
    const unlisten = history.listen((location: Location) => {
      locationObserver.current.notify(location.search);
    });

    // unsubscribe when the component unmounts
    return unlisten;
  }, []);

  const contextValue: FiltersContextValue = { 
    locationObserver,
    history, 
    filterRegistry: filterRegistry.current 
  };

  return (
    <filtersContext.Provider value={contextValue}>
      {children}
    </filtersContext.Provider>
  );
}


export default FiltersProvider;