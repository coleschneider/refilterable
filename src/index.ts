import FiltersProvider from "./FiltersProvider";
import createFilter from "./createFilter";
import useFilter from "./useFilter";
import useReset from "./useReset";

export { useReset, useFilter, createFilter, FiltersProvider };
