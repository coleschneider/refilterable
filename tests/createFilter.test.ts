import { createFilter } from '../src';
import { isFilterObject } from '../src/utils/types';


describe('createFilter', () => {
  it('should throw an error if the param name is not specified', () => {
    expect(() => {
      // @ts-ignore
      createFilter();
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      // @ts-ignore
      createFilter('');
    }).toThrowErrorMatchingSnapshot();
  });

  it('should pass the url param name to the configuration', () => {
    expect(createFilter('foo').paramName).toBe('foo');
  });

  it('should provide default parse, format, and validate methods', () => {
    const filter = createFilter('foo');

    expect(filter.parse('foo')).toBe('foo');
    expect(filter.format(123)).toBe('123');
    expect(filter.validate('foo', filter.parse)).toBe(true);
  });

  test('resetValue should fall back to defaultValue', () => {
    const filter = createFilter('foo', { defaultValue: "bar" });

    expect(filter.resetValue).toBe(filter.defaultValue);
    expect(filter.resetValue).toBe("bar");
  });

  it('should allow the user to override parse, format, and validate methods', () => {
    const parse = jest.fn();
    const format = jest.fn();
    const validate = jest.fn().mockImplementation((input, parse) => parse(input) && true);
    const filter = createFilter('foo', { parse, format, validate });

    filter.parse('foo');
    expect(parse).toHaveBeenCalledWith('foo');
    filter.format(123);
    expect(format).toHaveBeenCalledWith(123);
    filter.validate('foo', parse);
    expect(validate).toHaveBeenCalledWith('foo', parse);
    expect(parse).toHaveBeenLastCalledWith('foo');
  });

  it('should add a symbol specifying that the filter configuration was created using createFilter', () => {
    const filterObject = createFilter('foo');
    const notFilterObject = JSON.parse(JSON.stringify(filterObject));
    expect(isFilterObject(filterObject)).toBe(true);
    expect(isFilterObject(notFilterObject)).toBe(false);
  });
});
