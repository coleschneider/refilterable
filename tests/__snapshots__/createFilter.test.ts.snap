// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createFilter should throw an error if the param name is not specified 1`] = `"refilterable: param name cannot be empty"`;

exports[`createFilter should throw an error if the param name is not specified 2`] = `"refilterable: param name cannot be empty"`;
